File: 3sum.rb
#!/usr/bin/env ruby
#
# https://leetcode.com/problems/3sum/description/
#

def three_sum nums
  return [[0, 0, 0]] if nums.uniq == 0 and nums.length > 2
  return [] if nums.length < 3
  nums.sort!
  result = []

  m = 0
  while m < nums.length - 2
    i = m + 1
    j = nums.length - 1
 
    a = nums[m]
    while i < j
      b = nums[i]
      c = nums[j]

      if (a + b + c) == 0
        result << [a, b, c]
        j -= 1
        while j > i and nums[j] == nums[j + 1]
          j -= 1
        end
        i += 1
        while i < j and nums[i] == nums[i - 1]
          i += 1
        end
      elsif (a + b + c) > 0
        j -= 1
        while j > i and nums[j] == nums[j + 1]
          j -= 1
        end
      else
        i += 1
        while i < j and nums[i] == nums[i - 1]
          i += 1
        end
      end
    end
  
    m += 1
    while m < nums.length - 2 and nums[m] == nums[m - 1]
      m += 1
    end
  end

  return result
end

a = three_sum [-4,-2,-2,-2,0,1,2,2,2,3,3,4,4,6,6]
a.each do |i|
  i.each {|n| print "#{n}, "}
  puts
end

File: 3sum.c
#include <stdio.h>
#include <stdlib.h>

int *compar(const void *)

void three_sum(int nums[], int n) {
  int m = 0;
  while (m < n - 3) {
    int i = m + 1;
    int j = n - 1;

    while (i < j) {
      int result = nums[m] + nums[i] + nums[j];
      if (result == 0) {
        printf("%d = %d + %d + %d", result, nums[m], nums[i], nums[j]);
        i++;
        j--;
      } else if (result > 0) {
        j--;
      } else {
        i++;
      }

      m++;
    }
  }
}
int main(int argc, char *argv[]) {
  int nums[] = {-2, 0, 2}
  three_sum()
  return 0;
}

